command/hcl2_upgrade.go

```diff
@@ -0,0 +1,294 @@
+package command
+
+import (
+	"bytes"
+	"context"
+	"fmt"
+	"io"
+	"os"
+	"path/filepath"
+	"sort"
+	"strings"
+	texttemplate "text/template"
+
+	"github.com/hashicorp/hcl/v2/hclwrite"
+	hcl2shim "github.com/hashicorp/packer/hcl2template/shim"
+	"github.com/hashicorp/packer/template"
+	"github.com/posener/complete"
+	"github.com/zclconf/go-cty/cty"
+)
+
+type HCL2UpgradeCommand struct {
+	Meta
+}
+
+func (c *HCL2UpgradeCommand) Run(args []string) int {
+	ctx, cleanup := handleTermInterrupt(c.Ui)
+	defer cleanup()
+
+	cfg, ret := c.ParseArgs(args)
+	if ret != 0 {
+		return ret
+	}
+
+	return c.RunContext(ctx, cfg)
+}
+
+func (c *HCL2UpgradeCommand) ParseArgs(args []string) (*HCL2UpgradeArgs, int) {
+	var cfg HCL2UpgradeArgs
+	flags := c.Meta.FlagSet("hcl2_upgrade", FlagSetNone)
+	flags.Usage = func() { c.Ui.Say(c.Help()) }
+	cfg.AddFlagSets(flags)
+	if err := flags.Parse(args); err != nil {
+		return &cfg, 1
+	}
+	args = flags.Args()
+	if len(args) != 1 {
+		flags.Usage()
+		return &cfg, 1
+	}
+	cfg.Path = args[0]
+	if cfg.OutputFile == "" {
+		cfg.OutputFile = cfg.Path + ".pkr.hcl"
+	}
+	return &cfg, 0
+}
+
+func (c *HCL2UpgradeCommand) RunContext(buildCtx context.Context, cla *HCL2UpgradeArgs) int {
+
+	out := &bytes.Buffer{}
+	var output io.Writer
+	if err := os.MkdirAll(filepath.Dir(cla.OutputFile), 0); err != nil {
+		c.Ui.Error(fmt.Sprintf("Failed to create output directory: %v", err))
+		return 1
+	}
+	if f, err := os.Create(cla.OutputFile); err == nil {
+		output = f
+		defer f.Close()
+	} else {
+		c.Ui.Error(fmt.Sprintf("Failed to create output file: %v", err))
+		return 1
+	}
+
+	hdl, ret := c.GetConfigFromJSON(&cla.MetaArgs)
+	if ret != 0 {
+		return ret
+	}
+
+	core := hdl.(*CoreWrapper).Core
+	if err := core.Initialize(); err != nil {
+		c.Ui.Error(fmt.Sprintf("Initialization error: %v", err))
+		return 1
+	}
+	tpl := core.Template
+
+	// Output variables section
+
+	variables := []*template.Variable{}
+	{
+		// sort variables to avoid map's randomness
+
+		for _, variable := range tpl.Variables {
+			variables = append(variables, variable)
+		}
+		sort.Slice(variables, func(i, j int) bool {
+			return variables[i].Key < variables[j].Key
+		})
+	}
+
+	for _, variable := range variables {
+		variablesContent := hclwrite.NewEmptyFile()
+		variablesBody := variablesContent.Body()
+
+		variableBody := variablesBody.AppendNewBlock("variable", []string{variable.Key}).Body()
+
+		if variable.Default != "" || !variable.Required {
+			variableBody.SetAttributeValue("default", hcl2shim.HCL2ValueFromConfigValue(variable.Default))
+		}
+		if isSensitiveVariable(variable.Key, tpl.SensitiveVariables) {
+			variableBody.SetAttributeValue("sensitive", cty.BoolVal(true))
+		}
+		variablesBody.AppendNewline()
+		out.Write(transposeTemplatingCalls(variablesContent.Bytes()))
+	}
+
+	fmt.Fprintln(out, `# "timestamp" template function replacement`)
+	fmt.Fprintln(out, `locals { timestamp = regex_replace(timestamp(), "[- TZ:]", "") }`)
```

Comment:
This is in all upgraded templates to avoid confusion regarding the `timestamp`?

Reply:
yes

Reply:
So that all people that use the timestamp function have it.

We could also make the code aware of the fact that the timestamp function was called but this was simpler for now.